### PR TITLE
Poison import if Qiskit 1.0+ is detected

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This uses `importlib.metadata` to search for installations of Qiskit 1.0 at runtime, and poison the import if one is detected.  The two packages are incompatible in a way that we cannot express to `pip` without having poisoned the `qiskit-terra` package as a whole, which is something we're less willing to do in order to keep it easier to install old versions of Qiskit-dependent software, and to keep the Qiskit 1.0 packaging story clean for the coming years.

It is not generally expected that a user specifically attempting to install Qiskit 0.45/0.46 will accidentally install Qiskit 1.0 as well. The intention of this poisoning in the "stable" branch is so that the case that a user installs Qiskit 1.0, but either also includes an old package with a dependency purely on `qiskit-terra`, or runs a subsequent `pip install` command whose version resolution causes `qiskit-terra` to become installed.  In both these cases, it could be the case that the `import qiskit` command finds the `qiskit-terra` `__init__.py` file, and so any error we also place in Qiskit 1.0 would have been skipped.

### Details and comments

I will say that I don't love the idea of poisoning a package by any stretch, but given all the problems we've seen where people have inadvertently been installing broken environments while trying to test Qiskit 1.0, I think this is the pragmatic solution.

The only case of false-positive poisoning I've been able to find is _very_ developer-centric:
1. a main-branch version of Qiskit 1.0 has been installed in editable mode
2. Qiskit 1.0 is uninstalled
3. `qiskit-terra` 0.45.x (this commit) is installed by any means, and the `qiskit` metapackage is either not installed, or installed but not editable in a way that overwrites the installations of step 1
4. the Python session can see the `qiskit.egg-info` directory that gets dropped by `pip install -e .` from step 1 _before_ any `dist-info` or `egg-info` directory of the old qiskit metapackage in the meta path (such as if the session is launched from the repo root).

If all those criteria are met, it's possible for an incompatible `qiskit.egg-info` to be in the _meta path_ without the actual Qiskit 1.0 code being in the _import path_, in which case the package will emit a false positive import error. *edit*: the only way I can think to get a false negative is if somebody manually deletes stuff out of `site-packages` without using `pip uninstall`, and that's not something we need to guard against.

That possibility was enough to scare me, so I threw in a `QISKIT_SUPPRESS_1_0_IMPORT_ERROR` environment variable, that can be set to `1` to override the error.

See #11618 for the companion PR on main.

### CODE REVIEW

**Comment by mtreinish** [comment]
How does this relate to: https://github.com/Qiskit/qiskit/pull/11230 the plan we were going with before was to warn on import

**Comment by jakelishman** [comment]
It's largely orthogonal to #11230 (which I'd forgotten about) - this PR fails the import if a `qiskit` distribution is installed with its major version non-zero, that PR raises a warning if the corresponding `qiskit` distribution isn't installed with 0.46.  Both are valid messages, and I think both should likely exist.

**Comment by jakelishman** [comment]
I've marked this as "on hold", because we're planning to move ahead with it, but to include a link to a migration guide with more detail in the error message first.

**Comment by jakelishman** [comment]
I've added a shortlink to the message which will work soon, but is blocked on Qiskit/documentation#710 being deployed.

**Comment by jakelishman** [comment]
Qiskit/documentation#710 should be merged and live now before any release would actually land.

File: releasenotes/notes/qiskit-1.0-incompatibility-9ad22472837837d8.yaml
Comment (mtreinish) [inline releasenotes/notes/qiskit-1.0-incompatibility-9ad22472837837d8.yaml]: I think maybe reinforcing that you can do this prior to qiskit 1.0 a bit more would be good. When I first read this I kind of got the impression that this was pushing for Qiskit 1.0 support. Maybe something like:

```suggestion
    now simply called "Qiskit"). As long as your library is compatible with qiskit/qiskit-terra >=0.44.0 you can
    substitute ``qiskit-terra`` with ``qiskit`` without issue (even if your library isn't ready for Qiskit 1.0 yet).
```

Although I'm not a huge fan of this suggestion either.

File: releasenotes/notes/qiskit-1.0-incompatibility-9ad22472837837d8.yaml
Comment (mtreinish) [inline releasenotes/notes/qiskit-1.0-incompatibility-9ad22472837837d8.yaml]: Do you want to repeat the link from the error message here too?

File: qiskit/__init__.py
Comment (mtreinish) [inline qiskit/__init__.py:30]: Do you think this should be documented anywhere? We do have: https://docs.quantum.ibm.com/start/configure-qiskit-local#environment-variables for env variables, but this seems esoteric and dangerous enough that it's probably not worth putting there.

File: qiskit/__init__.py
Comment (mtreinish) [inline qiskit/__init__.py]:
```suggestion
            " dependencies between Qiskit < 1.0.0 and >= 1.0.0"
```

File: qiskit/__init__.py
Comment (mtreinish) [inline qiskit/__init__.py]:
```suggestion
            "Qiskit is installed in an invalid environment that has both Qiskit >=1.0.0"
```

**Review by mtreinish — COMMENTED** [review]
This looks fine to me, I agree it's not a great situation but raising a clear error message that the install is corrupted is the lesser evil in this case. I just had a few small inline suggestions on wording then I think this is good to merge

File: qiskit/__init__.py
Comment (jakelishman) [reply qiskit/__init__.py:30]: It's at the very very bottom of https://qisk.it/packaging-1-0, and in some ways I was deliberately trying to obfuscate it because I don't want people to use it.  It's only intended as a relief valve if I've made a mistake in this PR.

I thought people who know enough to use it safely are likely to check the source code to see what checks it's doing before raising the error, in which case they'll see the envvar.

File: releasenotes/notes/qiskit-1.0-incompatibility-9ad22472837837d8.yaml
Comment (jakelishman) [reply releasenotes/notes/qiskit-1.0-incompatibility-9ad22472837837d8.yaml]: I think I left the blog post in as a placeholder.  I've put in proper links to the migration guide in 8ad7cf597d.

File: releasenotes/notes/qiskit-1.0-incompatibility-9ad22472837837d8.yaml
Comment (jakelishman) [reply releasenotes/notes/qiskit-1.0-incompatibility-9ad22472837837d8.yaml]: Added in 8ad7cf597d.

File: qiskit/__init__.py
Comment (jakelishman) [reply qiskit/__init__.py]: Done in 6df921fc68.

File: qiskit/__init__.py
Comment (jakelishman) [reply qiskit/__init__.py]: Done in 6df921fc68.

**Review by mtreinish — APPROVED** [review]
LGTM now, thanks for the fast update. We'll have to bump the version numbers to prepare a 0.45.3 release but at least the release notes are handled with this.

**Pull request merged**
